### PR TITLE
Moved all hooks-triggering calls inside transactions

### PR DESCRIPTION
**Pull request by oreilles**
TLDR: Throwing errors in any hook should rollback transactions

Description of my use case:
- Users create an item in collection A, which includes a CSV file containing data that will have to be inserted in collection B. Item A must be created only if the CSV content is valid, so I'm forced to use a `before` hook, which will lead problems because:
- To accomplish that, I must parse the file and insert its result in the database, without knowing if the user has valid credentials for collection A (because the hook is called before creation), or if the rest of the payload is valid - a potentially high workload that should be done only if item A's payload and credential have been validated
- Every item inserted in collection B (from the CSV file) must hold a foreign key reference to the item in collection A they come from. To acomplish that, I must:
  - Insert them first with an empty foreign key reference to A, since item A has not been created yet. Meaning that I can't have a `NOT NULL` constraint on that field if I wanted to. 
  - Return all their ids from the database operation with `knex.returning('id')`, [which is only available in PostgreSQL, MSSQL, and Oracle](http://knexjs.org/#Builder-returning) so if I were to use use MySQL I'd be out of luck
  - Put those ids in the payload so that they're available in another `item.create` hook
  - Now that item A is created and that I know its id, update all items with an expansive `update B set id_a = <item_a_pk> where id in (<very long list of all the imported items ids>)` query in the `item.create` hook - hoping that item A haven't been deleted in the meantime and that there haven't been any issue with the database or Directus server since I'm not in the same transaction anymore.

Conclusion:
- Waste of server resource in case item A's payload is invalid
- Prevent having item B -> item A NOT NULL foreign key constraint
- Huge database overhead
- Needs two hooks instead of one
- Actually impossible in vendors that doesn't support `.returning()` like MySQL
- All of this can be solved by making the `items.create` hook part of the transaction

This is just my use case but more generally, it's probably expected by users that the success of their CRUD operation depends on the success of all of their hooks, especially since they have access to the payload and database connection. If users need non-blocking, fail-safe hooks after a CRUD operation, the webhooks are much more appropriate.

Changes:
- Moved `item.create` hooks inside the transaction
- Moved `item.update` hooks inside the transaction
- Moved `item.delete` hooks inside the transaction
- Moved `item.update.before` hooks inside the transaction to follow `item.create.before` behaviour
- Moved `item.delete.before` hooks inside the transaction to follow `item.create.before` behaviour

Things I also noticed in the hook system and may need to be adressed:
- `updateField` and `deleteField` operations are not wrapped in transactions, which means any problem in that code or in the database during execution will probably result in a corrupted state
- All CRUD operations in `items.ts`, `fields.ts`, `authentification.ts` rely on an `AuthorisationService` which ~~never~~ most of the time doesn't share the same transaction as the rest of the operations, which once again could lead to some vulnerabilities

Thoughts ?

### CODE REVIEW

**Comment by rijkvanzanten** [comment]
The .before hooks should definitely be in the transaction (that was intended), but i'm not too sure about the "regular" ones. Those are more meant to be used for an async operation after the fact, that isn't supposed to block the operation or cancel any insertion 🤔 (like sending an email once an item is created). I don't think we'd want that in transaction. That's also the reason why they were in a `emitAsyncSafe`, which you just removed completely 👀 

> This is just my use case but more generally, it's probably expected by users that the success of their CRUD operation depends on the success of all of their hooks, especially since they have access to the payload and database connection. If users need non-blocking, fail-safe hooks after a CRUD operation, the webhooks are much more appropriate.

Webhooks are just a thing GUI shell around regular "non .before" hooks, and can't be considered a replacement for these hooks, as they're limited to specific functionality (by design)

Regarding the use case:

> Users create an item in collection A, which includes a CSV file containing data that will have to be inserted in collection B. 

Any reason why you wouldn't just add the nested items in a o2m field on collection A? Directus already has all the checks and transactions in place 🤔  

---

I think the real solution here is to add a new (third) hook that is a _blocking_ "after"

**Comment by Oreilles** [comment]
Then maybe there should be another hook that could be named `items.create.after` ? If we can do some work *before* as part of the transaction, we should be able to do the same *after* the item have been created. Even if it's just an email, you might want to be 100% sure that the code you wrote to send that email (or whatever) has completed in order for the transaction to complete.

**Comment by ijpatricio** [comment]
Hey @Oreilles ,

Did you get a good result with an O2M field??

**Comment by rijkvanzanten** [comment]
The `action` hooks should remain async and non-blocking as per the comments above. We can discuss the addition of a new "filter after" type event in a new feature request discussion 👍🏻
